controller/volume_controller.go

```diff
@@ -1141,10 +1142,120 @@ func (vc *VolumeController) ReconcileVolumeState(v *longhorn.Volume, es map[stri
 		if oldState != v.Status.State {
 			vc.eventRecorder.Eventf(v, v1.EventTypeNormal, EventReasonAttached, "volume %v has been attached to %v", v.Name, v.Status.CurrentNodeID)
 		}
+
+		keepLocalReplicaPolicy, err := vc.ds.GetSettingAsBool(types.SettingNameKeepLocalReplicaOnSameNodeAsEngine)
+		if err != nil {
+			return err
+		}
+		if keepLocalReplicaPolicy {
+			if err := vc.handleKeepLocalReplicaOnSameNodeAsEngine(v, e, rs); err != nil {
+				return err
+			}
+		}
 	}
 	return nil
 }
 
+func (vc *VolumeController) handleKeepLocalReplicaOnSameNodeAsEngine(v *longhorn.Volume, e *longhorn.Engine, rs map[string]*longhorn.Replica) error {
+	if e == nil {
+		return fmt.Errorf("BUG: handleKeepLocalReplicaWithEngine needs a valid engine")
+	}
+
+	if !vc.isSafeToModifyReplica(v, e, rs) {
+		return nil
+	}
+
+	// hasLocalReplica indicate whether there exist a replica on the same node as engine
+	hasLocalReplica := false
+	for _, r := range rs {
+		if r.Spec.NodeID == e.Spec.NodeID {
+			hasLocalReplica = true
+		}
+	}
+
+	if !hasLocalReplica {
+		if _, err := vc.addLocalReplicaOnSameNodeAsEngine(v, e, rs); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	// check if there are have extra healthy replicas
+	healthyCount := 0
+	for rName, mode := range e.Status.ReplicaModeMap {
```

Comment:
We can directly verify if a replica is healthy by checking `r.Spec.HealthyAt != "" && r.Spec.FailedAt  == ""`

Reply:
Thanks. Fixed